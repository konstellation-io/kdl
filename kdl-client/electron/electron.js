const electron = require('electron');
const app = electron.app;
const BrowserWindow = electron.BrowserWindow;

const path = require('path');
const isDev = require('electron-is-dev');
const { autoUpdater } = require('electron-updater');

const SPLASH_SCREEN_MIN_TIME = 2000;
const SPLASH_SCREEN_SIZE = {
  width: 600,
  height: 400,
};
const WINDOW_SIZE = {
  width: 900,
  height: 680,
};

let mainWindow;
let splashScreen;

function showSplashScreen() {
  splashScreen = new BrowserWindow({
    ...SPLASH_SCREEN_SIZE,
    frame: false,
    backgroundColor: '#060b15',
  });

  splashScreen.loadURL(
    `file://${path.join(__dirname, '../public/splash_screen.html')}`
  );
}

function closeSplashScreen() {
  splashScreen.close();
  splashScreen = null;
}

function createMainWindow() {
  let mainWindowReady = false;

  mainWindow = new BrowserWindow({
    ...WINDOW_SIZE,
    show: false,
    webPreferences: {
      nodeIntegration: true,
    },
    transparent: true,
    // frame: false,   // TODO: set this to true after creating a custom title bar
    titleBarStyle: 'hidden',
  });

  mainWindow.loadURL(
    isDev
      ? 'http://localhost:3000'
      : `file://${path.join(__dirname, '../build/index.html')}`
  );

  if (isDev) {
    mainWindow.webContents.openDevTools();
  }

  mainWindow.once('ready-to-show', () => {
    if (!mainWindow) {
      throw new Error('"mainWindow" is not defined');
    }

    mainWindowReady = true;
  });

  // This will wait at least for two seconds before showing the main window.
  const showMainWindowInterval = setInterval(() => {
    if (mainWindow && mainWindowReady) {
      clearInterval(showMainWindowInterval);

      closeSplashScreen();

      mainWindow.show();
      mainWindow.focus();
    }
  }, SPLASH_SCREEN_MIN_TIME);

  mainWindow.on('closed', () => (mainWindow = null));
}

function createWindow() {
  showSplashScreen();
  createMainWindow();
}

// when the app is loaded create a BrowserWindow and check for updates
app.on('ready', function () {
  createWindow();
  if (!isDev) autoUpdater.checkForUpdates();
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', () => {
  mainWindow && createWindow();
});

// when the update has been downloaded and is ready to be installed, notify the BrowserWindow
autoUpdater.on('update-downloaded', () => {
  mainWindow.webContents.send('updateReady');
});

// when receiving a quitAndInstall signal, quit and install the new version ;)
electron.ipcMain.on('quitAndInstall', () => {
  autoUpdater.quitAndInstall();
});
