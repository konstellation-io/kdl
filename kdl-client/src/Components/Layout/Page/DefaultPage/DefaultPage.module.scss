@import 'Styles/colors';
@import 'Styles/variables';
@import 'Styles/mixins';

$sidebar-width: 40 * $grid-unit;

.container {
  position: relative;
  display: flex;
  width: 100%;
  height: 100%;
  background-color: palette(base, 700);
}

.actions div {
  padding: 0;
  flex-direction: row;
}

.sidebar {
  width: $sidebar-width;
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  background-color: palette(base, 500);
  padding: 10 * $grid-unit 5 * $grid-unit 0 3 * $grid-unit;

  h1 {
    text-align: left;
  }

  .subtitle {
    margin-top: 3 * $grid-unit;
  }

  .bottom {
    margin-top: auto;
  }
}

.content {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  max-width: 1440px - $sidebar-width;
  background-color: palette(base, 800);
  overflow: auto;
}

.children {
  flex-grow: 1;
}
