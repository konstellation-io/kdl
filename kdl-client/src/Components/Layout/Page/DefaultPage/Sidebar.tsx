import React, { useRef } from 'react';

import styles from './DefaultPage.module.scss';

export let bottomRef: React.MutableRefObject<HTMLDivElement | null>;
export let topRef: React.MutableRefObject<HTMLDivElement | null>;

type Props = {
  title: string;
  subtitle?: string;
};
function Sidebar({ title, subtitle }: Props) {
  bottomRef = useRef<HTMLDivElement>(null);
  topRef = useRef<HTMLDivElement>(null);

  return (
    <div className={styles.sidebar}>
      <h1>{title}</h1>
      {subtitle && <h3 className={styles.subtitle}>{subtitle}</h3>}
      <div className={styles.top} ref={topRef} />
      <div className={styles.bottom} ref={bottomRef} />
    </div>
  );
}

export default Sidebar;
