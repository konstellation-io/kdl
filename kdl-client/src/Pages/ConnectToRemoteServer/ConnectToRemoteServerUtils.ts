import { CHECK } from 'kwc';

export function validateServerUrl(value: string, usedUrls: string[]) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(value),
    CHECK.isUrlValid(value),
    CHECK.isItemDuplicated(value, usedUrls, 'server URL'),
  ]);
}
