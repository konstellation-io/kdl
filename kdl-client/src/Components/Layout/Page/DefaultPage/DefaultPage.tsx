import React, { FC } from 'react';

import ActionsBar from 'Components/Layout/ActionsBar/ActionsBar';
import Sidebar from './Sidebar';
import styles from './DefaultPage.module.scss';

type Props = {
  title: string;
  subtitle?: string;
  children: JSX.Element;
  actions?: JSX.Element | JSX.Element[];
};
const DefaultPage: FC<Props> = ({ title, subtitle, actions, children }) => (
  <div className={styles.container}>
    <Sidebar title={title} subtitle={subtitle} />
    <div className={styles.content}>
      <div className={styles.children}>{children}</div>
      {actions && <ActionsBar className={styles.actions}>{actions}</ActionsBar>}
    </div>
  </div>
);

export default DefaultPage;
