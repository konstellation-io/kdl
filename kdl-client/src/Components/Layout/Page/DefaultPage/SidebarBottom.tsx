import React, { FC } from 'react';

import { bottomRef } from './Sidebar';
import { createPortal } from 'react-dom';
import styles from './DefaultPage.module.scss';

type Props = {
  children: JSX.Element;
};
const SidebarBottom: FC<Props> = ({ children }) => (
  <div className={styles.container}>
    {bottomRef?.current && createPortal(children, bottomRef.current)}
  </div>
);

export default SidebarBottom;
