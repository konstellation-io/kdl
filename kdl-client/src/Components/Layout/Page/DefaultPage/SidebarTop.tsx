import React, { FC } from 'react';

import { createPortal } from 'react-dom';
import styles from './DefaultPage.module.scss';
import { topRef } from './Sidebar';

type Props = {
  children: JSX.Element;
};
const SidebarTop: FC<Props> = ({ children }) => (
  <div className={styles.container}>
    {topRef?.current && createPortal(children, topRef.current)}
  </div>
);

export default SidebarTop;
