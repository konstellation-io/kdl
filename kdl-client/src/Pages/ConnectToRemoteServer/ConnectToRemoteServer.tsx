import { Button, CHECK, TextInput } from 'kwc';
import ROUTE, { buildRoute } from 'Constants/routes';
import React, { useCallback, useEffect, useState } from 'react';
import StatusCircle, {
  States,
} from 'Components/LottieShapes/StatusCircle/StatusCircle';

import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import IconLink from '@material-ui/icons/Link';
import SidebarBottom from 'Components/Layout/Page/DefaultPage/SidebarBottom';
import { ipcRenderer } from 'electron';
import styles from './ConnectToRemoteServer.module.scss';
import { useForm } from 'react-hook-form';
import { useHistory } from 'react-router-dom';
import useServers from 'Hooks/useServers';

type ConnectToRemoteServerResponse = {
  serverId?: string;
  success: boolean;
  error?: string;
};

enum ConnectionState {
  CONNECTING = 'CONNECTING',
  OK = 'OK',
  ERROR = 'ERROR',
}

function getStateCircle(state: ConnectionState) {
  switch (state) {
    case ConnectionState.OK:
      return <StatusCircle animation={States.SUCCESS} label="CONNECTED" />;
    case ConnectionState.ERROR:
      return <StatusCircle animation={States.ERROR} label="ERROR" />;
    default:
      return <StatusCircle label="CONNECTING" />;
  }
}

type FormData = {
  serverUrl: string;
};

function ConnectToRemoteServer() {
  const [connectionState, setConnectionState] = useState<ConnectionState>();
  const { servers } = useServers();
  const history = useHistory();

  const {
    handleSubmit,
    setValue,
    unregister,
    register,
    errors,
    setError,
    clearErrors,
  } = useForm<FormData>({ defaultValues: { serverUrl: '' } });

  const validateUrl = useCallback(
    (value: string) => {
      const serverUrls = servers.map((server) => server.url || '');

      return CHECK.getValidationError([
        CHECK.isFieldNotEmpty(value),
        CHECK.isDomainValid(value),
        CHECK.isItemDuplicated(value, serverUrls, 'server URL'),
      ]);
    },
    [servers]
  );

  useEffect(() => {
    register('serverUrl', { validate: validateUrl });
    return () => unregister('serverUrl');
  }, [register, unregister, setValue, validateUrl]);

  useEffect(() => {
    const onConnectToRemoteServerReply = (
      _: unknown,
      { error, success, serverId }: ConnectToRemoteServerResponse
    ) => {
      if (success && !error) {
        setConnectionState(ConnectionState.OK);

        setTimeout(() => {
          history.push(buildRoute.server(ROUTE.SERVER_LOGIN, serverId || ''));
        }, 2000);
      } else {
        setConnectionState(ConnectionState.ERROR);
        setError('serverUrl', { message: error });
      }
    };

    ipcRenderer.on('connectToRemoteServerReply', onConnectToRemoteServerReply);

    return () => {
      ipcRenderer.removeListener(
        'connectToRemoteServerReply',
        onConnectToRemoteServerReply
      );
    };
    // We want to execute this on on component mount/unmount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  async function onSubmit({ serverUrl }: FormData) {
    ipcRenderer.send('connectToRemoteServer', {
      url: serverUrl,
    });
    setConnectionState(ConnectionState.CONNECTING);
  }

  // Show actions only not connecting or there was a connection error
  let actions = undefined;
  if (
    connectionState === undefined ||
    connectionState === ConnectionState.ERROR
  ) {
    actions = [
      <Button
        label="CANCEL"
        className={styles.cancelButton}
        key="button-2"
        onClick={() => history.goBack()}
      />,
      <Button
        label="CONNECT"
        className={styles.connectButton}
        onClick={handleSubmit(onSubmit)}
        key="button-1"
        primary
      />,
    ];
  }

  return (
    <DefaultPage
      title="Connect to a Remote Server"
      subtitle="Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum."
      actions={actions}
    >
      <div className={styles.container}>
        <p className={styles.title}>How to connect to a Server</p>
        <p className={styles.subtitle}>
          In hac habitasse platea dictumst. Vivamus adipiscing fermentum quam
          volutpat aliquam. Integer et elit eget elit facilisis tristique. Nam
          vel iaculis mauris. Sed ullamcorper tellus erat, non ultrices sem
          tincidunt euismod. Fusce rhoncus porttitor velit, eu bibendum nibh
          aliquet vel. Fusce lorem leo, vehicula at nibh quis, facilisis
          accumsan turpis.
        </p>
        <div className={styles.formServerUrl}>
          <TextInput
            label="server url"
            onChange={(value: string) => {
              errors.serverUrl && clearErrors('serverUrl');
              setValue('serverUrl', value);
            }}
            onEnterKeyPress={handleSubmit(onSubmit)}
            Icon={IconLink}
            error={errors.serverUrl?.message}
            disabled={
              connectionState &&
              [ConnectionState.CONNECTING, ConnectionState.OK].includes(
                connectionState
              )
            }
            showClearButton
          />
        </div>
        {connectionState && (
          <SidebarBottom>
            <div className={styles.loader}>
              {getStateCircle(connectionState)}
            </div>
          </SidebarBottom>
        )}
      </div>
    </DefaultPage>
  );
}

export default ConnectToRemoteServer;
