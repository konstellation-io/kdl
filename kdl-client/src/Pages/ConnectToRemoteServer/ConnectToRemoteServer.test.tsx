import { Button } from 'kwc';
import ConnectToRemoteServer from './ConnectToRemoteServer';
import DefaultPage from 'Components/Layout/Page/DefaultPage/DefaultPage';
import React from 'react';
import SidebarBottom from 'Components/Layout/Page/DefaultPage/SidebarBottom';
import { ipcRenderer } from 'electron';
import { shallow } from 'enzyme';

const mockGoBack = jest.fn();
jest.mock('react-router-dom', () => ({
  useHistory: jest.fn(() => ({
    goBack: mockGoBack,
  })),
}));

jest.mock('electron', () => ({
  ipcRenderer: {
    send: jest.fn(),
  },
}));

// TODO: study how to test forms with useForm (act is causing infinite loops)

describe('ConnectToRemoteServer page', () => {
  let component: any;

  beforeEach(() => {
    component = shallow(<ConnectToRemoteServer />);
  });

  test('Component match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('show right texts', () => {
    expect(component.contains('How to connect to a Server')).toBeTruthy();
  });

  test('Initially is not connecting', () => {
    expect(component.find(SidebarBottom).isEmptyRender()).toBeTruthy();
  });

  test('back action', () => {
    component.find(DefaultPage).dive().find(Button).first().simulate('click');

    expect(mockGoBack).toHaveBeenCalledTimes(1);
  });

  test('action button', () => {
    component.find(DefaultPage).dive().find(Button).last().simulate('click');

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);

    // While checking, status circle is shown at sidebar bottom and buttons are hidden
    expect(component.find(SidebarBottom).isEmptyRender()).toBeFalsy();
    expect(
      component.find(DefaultPage).dive().find(Button).isEmptyRender()
    ).toBeTruthy();
  });
});
