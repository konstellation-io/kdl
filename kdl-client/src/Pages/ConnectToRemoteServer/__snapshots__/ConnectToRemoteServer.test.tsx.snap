// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectToRemoteServer page Component match snapshot 1`] = `
<DefaultPage
  actions={
    Array [
      <P
        className="cancelButton"
        label="CANCEL"
        onClick={[Function]}
      />,
      <P
        className="connectButton"
        label="CONNECT"
        onClick={[Function]}
        primary={true}
      />,
    ]
  }
  subtitle="By connecting to a Remote Server, you can access projects located in some machine or cluster shared among different users."
  title="Connect to a Remote Server"
>
  <div
    className="container"
  >
    <p
      className="title"
    >
      How to connect to a Server
    </p>
    <p
      className="subtitle"
    >
      Before trying to connect to a Server, make sure you have access privileges. A Server administrator must add you account to the users list. You can add a Remote Server without access provileges but you will not be able to sign in or work with projects.
    </p>
    <div
      className="formServerUrl"
    >
      <tt
        Icon={
          Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "LinkIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          }
        }
        label="server url"
        onChange={[Function]}
        onEnterKeyPress={[Function]}
        showClearButton={true}
      />
    </div>
  </div>
</DefaultPage>
`;
